nym-vpn-core/crates/nym-vpn-lib-types/src/tunnel_state.rs

```diff
@@ -198,6 +205,117 @@ pub enum ErrorStateReason {
+                            Self::BandwidthExceeded
+                        }
+                        Some(id) if id.contains(SUBSCRIPTION_EXPIRED_MESSAGE_ID) => {
+                            Self::SubscriptionExpired
+                        }
+                        _ => Self::Api(e.clone().message),
+                    };
+                }
+                if let Some(err) = failed
+                    .iter()
+                    .find(|e| matches!(e, RequestZkNymErrorReason::UnexpectedVpnApiResponse { .. }))
+                {
+                    return Self::Api(err.to_string());
+                }
+                Self::Internal(failed.into_iter().map(|e| e.to_string()).collect())
+            }
+            ErrorStateReason::Firewall => Self::Firewall,
+            ErrorStateReason::TunDevice
+            | ErrorStateReason::TunnelProvider
+            | ErrorStateReason::DuplicateTunFd
+            | ErrorStateReason::Dns => Self::Internal(value.to_string()),
+            ErrorStateReason::Internal(message) => Self::Internal(message),
+            ErrorStateReason::Routing => Self::Routing,
+            ErrorStateReason::ResolveGatewayAddrs => Self::Dns(value.to_string()),
+            ErrorStateReason::StartLocalDnsResolver => Self::Dns(value.to_string()),
+        }
+    }
+}
+
+impl fmt::Display for ErrorStateReason {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ErrorStateReason::Firewall => write!(f, "Firewall"),
+            ErrorStateReason::Routing => write!(f, "Routing"),
+            ErrorStateReason::Dns => write!(f, "Dns"),
+            ErrorStateReason::TunDevice => write!(f, "TunnelDevice"),
+            ErrorStateReason::TunnelProvider => write!(f, "TunnelProvider"),
+            ErrorStateReason::SameEntryAndExitGateway => write!(f, "SameEntryAndExitGateway"),
+            ErrorStateReason::InvalidEntryGatewayCountry => write!(f, "InvalidEntryGatewayCountry"),
+            ErrorStateReason::InvalidExitGatewayCountry => write!(f, "InvalidExitGatewayCountry"),
+            ErrorStateReason::BadBandwidthIncrease => write!(f, "BadBandwidthIncrease"),
+            ErrorStateReason::DuplicateTunFd => write!(f, "DuplicateTunFd"),
+            ErrorStateReason::SyncAccount(_) => write!(f, "SyncAccount"),
+            ErrorStateReason::SyncDevice(_) => write!(f, "SyncDevice"),
+            ErrorStateReason::RegisterDevice(_) => write!(f, "RequestZkNym"),
+            ErrorStateReason::RequestZkNym(_) => write!(f, "InvalidExitGatewayCountry"),
+            ErrorStateReason::RequestZkNymBundle { .. } => write!(f, "RequestZkNymBundle "),
+            ErrorStateReason::Internal(_) => write!(f, "Internal"),
+            ErrorStateReason::ResolveGatewayAddrs => write!(f, "ResolveGatewayAddrs"),
+            ErrorStateReason::StartLocalDnsResolver => write!(f, "StartLocalDnsResolver"),
+        }
+    }
+}
```

Comment:
suggestion: to automatically implement `Display` on an enum you can use [strum](https://docs.rs/strum/latest/strum/derive.Display.html) crate

```
#[derive(Debug, Clone, Eq, PartialEq, strum_macros::Display)]
pub enum ErrorStateReason {
  // [...]
}
```